### PR TITLE
[JUJU-2331] Add cli command for update-secret-backend and associated facade apis

### PR DESCRIPTION
**Pull request by wallyworld**
We have add, list, remove... this PR adds a command to update secret backends.
The changes include validating that the new config is ok (ping the backend). The PR also updates the add api to use ping.

When updating backend config, values can be reset. The PR adds default value support for the provider config, so that if a value is reset and a default exists, use that.
To disable token rotation, set the new value to 0.
A backend can also be renamed using name=foo as an option.

We also add a validation check to model-config so that setting a model's secret backend checks that the backend exists.

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

start a vault server
add a secret backend with an invalid endpoint
```
juju add-secret-backend myvault vault --config ~/vaultcfg.yaml
ERROR backend not reachable: Get "http://10.0.0.77:8201/v1/sys/health?drsecondarycode=299&performancestandbycode=299&sealedcode=299&standbycode=299&uninitcode=299": EOF
```
add the backend with valid config

after each update, run `juju secret-backends --format yaml` to check results

rename the backend
`juju update-secret-backend myvault name=myvault2`

add a token rotation
`juju update-secret-backend myvault2 token-rotate=2m`

remove token rotation
`juju update-secret-backend myvault2 token-rotate=0`

update config
`juju update-secret-backend myvault2 namespace=foo`

reset config
`juju update-secret-backend myvault2 --reset namespace`

update config with an invalid endpoint and check error (as for add)

check model-config validation
```juju model-config secret-backend=foo
ERROR secret backend "foo" not found (not found)
```
`juju model-config secret-backend=myvault2`
`juju model-config secret-backend=auto`

### CODE REVIEW

File: cmd/juju/secretbackends/add.go [hunk 1]
```diff
@@ -152,28 +160,36 @@ func (c *addSecretBackendCommand) readFile(ctx *cmd.Context) (map[string]interfa
...
 	const tokenRotate = "token-rotate"
 	var tokenRotateInterval *time.Duration
 	tokenRotateStr, ok := attrs[tokenRotate]
 	if ok {
 		delete(attrs, tokenRotate)
 		rotateInterval, err := time.ParseDuration(fmt.Sprintf("%s", tokenRotateStr))
-		if err != nil || (rotateInterval/time.Second) < 60 {
-			return errors.NotValidf("token rotate interval %q", tokenRotateStr)
+		intervalSecs := rotateInterval / time.Second
+		if err != nil || intervalSecs < 60 && (!zeroAllowed || intervalSecs != 0) {
```
Comment (hpidcock) [inline cmd/juju/secretbackends/add.go:171]: these should be seperate error messages (one for less than 60s, one for zero) and actually expose the time.ParseDuration error.

File: state/secretbackends.go [hunk 2]
```diff
@@ -116,6 +125,62 @@ func (s *secretBackendsStorage) toSecretBackend(doc *secretBackendDoc) *secrets.
...
 }
 
+// UpdateSecretBackend updates a new secret backend.
+func (s *secretBackendsStorage) UpdateSecretBackend(p UpdateSecretBackendParams) error {
+	secretBackendsColl, closer := s.st.db().GetCollection(secretBackendsC)
+	defer closer()
+
+	var doc secretBackendDoc
+	err := secretBackendsColl.FindId(p.ID).One(&doc)
+	if err == mgo.ErrNotFound {
```
Comment (hpidcock) [inline state/secretbackends.go:135]: should really use errors.Is rather than relying on equality.

**Review by hpidcock — APPROVED** [review]

File: state/secretbackends.go (hunk 2 above)
Comment (wallyworld) [reply state/secretbackends.go:135]: we use this in 105 other places - the mgo api returns ErrNotFound directly

**Pull request merged**